### PR TITLE
c++ compatible using abseil-cpp

### PR DESCRIPTION
**Pull request by hp03**
replace `std::variant` with `absl::variant`;
replace `std::any` with `absl::any`;
replace `std::string_view` with `absl::string_view`;
replace `std::unordered_map` with `absl::flat_hash_map`;

replace `if constexpr` with SFINAE template functions;
replace `auto [a, b] = ` with multiple auto statements;

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "abseil-cpp"]
```
Comment (Superjomn) [inline .gitmodules]: Use cmake instead.

Avoid submodule to make maintain easier.

File: cinn/backends/generated_module1.cc [hunk 2]
```diff
@@ -1 +1,2 @@
-#include "cinn/backends/_generated_module1.cc"
+// TODO temporarily disable this file
```
Comment (Superjomn) [inline cinn/backends/generated_module1.cc]: 为啥会有重复 include ？

File: cinn/hlir/framework/graph_compiler.h [hunk 3]
```diff
@@ -9,7 +9,7 @@
 #include "cinn/backends/cuda_util.h"
 #include "cinn/common/macros.h"
 #include "cinn/hlir/framework/graph.h"
-#include "cinn/hlir/framework/instruction.h"
+#include "cinn/hlir/framework/instruction.h" // TODO !!!!!
```
Comment (Superjomn) [inline cinn/hlir/framework/graph_compiler.h]: ？

File: cinn/hlir/framework/instruction.h [hunk 4]
```diff
@@ -4,7 +4,7 @@
 #include <vector>
 
 #include "cinn/backends/cuda_util.h"
-#include "cinn/common/test_helper.h"
+// #include "cinn/common/test_helper.h"
```
Comment (Superjomn) [inline cinn/hlir/framework/instruction.h]: clean this line

File: cinn/hlir/framework/op.h [hunk 5]
```diff
@@ -202,6 +203,11 @@ bool OpValueType<ValueType>::Find(const Operator* op) const {
   return idx < data.size();
 }
 
+template <typename R, typename ...Args>
+inline auto makeOpFunction(R(*func)(Args...)) {
```
Comment (Superjomn) [inline cinn/hlir/framework/op.h]: MakeOpFunction

File: cinn/pybind/bind_utils.h [hunk 6]
```diff
@@ -22,15 +21,15 @@ using common::Type;
...
 using ir::ExprNode;
 
-template <class... Ts>
-struct Visitor : Ts... {
-  using Ts::operator()...;
-};
-
-template <class... Ts>
-Visitor(Ts...)->Visitor<Ts...>;
+// template <class... Ts>
```
Comment (Superjomn) [inline cinn/pybind/bind_utils.h]: clean dropped code

File: tools/auto_convert_structured_binding.awk [hunk 7]
```diff
@@ -0,0 +1,26 @@
+function convert(indent, line) {
+  match(line, /^(.*auto\s+&*)\[([^\]]*)\](.*)$/, arr) # certainly always match
+  split(arr[2], vars, /\s*,\s*/)
+  mixvar = "_"
+  for (vi in vars) {
+    mixvar = mixvar vars[vi] "_"
+  }
+  print arr[1] mixvar arr[3]
```
Comment (Superjomn) [inline tools/auto_convert_structured_binding.awk]: Remove this?

**Comment by haozech** [comment]
经测试此pr对性能无负面影响。
数据如下：
develop:
resnet18 - cudnn_off: 5.697 ms  5.734 ms
mobilenetV2 - cudnn_off: 3.420 ms 3.417 ms
efficientnet - cudnn_off: 6.680 ms  6.694 ms

resnet18 - cudnn_on: 2.263 ms  2.151 ms
mobilenetV2 - cudnn_on: 1.418 ms 1.421 ms
efficientnet - cudnn_on: 2.423 ms  2.426 ms

resnet18 - gpu_off: 37.472 ms  37.928 ms
mobilenetV2 - gpu_off: 19.375 ms 19.436 ms
efficientnet - gpu_off: 30.352 ms 29.489 ms

c++14_compitable branch:
resnet18 - cudnn_off: 5.719 ms
mobilenetV2 - cudnn_off: 3.436 ms
efficientnet - cudnn_off: 6.681 ms

resnet18 - cudnn_on: 2.082 ms
mobilenetV2 - cudnn_on: 1.402 ms
efficientnet - cudnn_on: 2.436 ms

resnet18 - gpu_off:  36.570 ms
mobilenetV2 - gpu_off: 19.191 ms
efficientnet - gpu_off: 30.618 ms

**Review by haozech — APPROVED** [review]
LGTM

**Pull request merged**